mobile/library/common/http/client.cc

```diff
@@ -406,54 +412,77 @@ void Client::DirectStream::saveLatestStreamIntel() {
 }
 
 void Client::DirectStream::saveFinalStreamIntel() {
+  if (!parent_.getStream(stream_handle_, ALLOW_ONLY_FOR_OPEN_STREAMS)) {
+    return;
+  }
   OptRef<RequestDecoder> request_decoder = requestDecoder();
-  if (!request_decoder || !parent_.getStream(stream_handle_, ALLOW_ONLY_FOR_OPEN_STREAMS)) {
+  if (!request_decoder) {
     return;
   }
   StreamInfo::setFinalStreamIntel(request_decoder->streamInfo(), parent_.dispatcher_.timeSource(),
                                   envoy_final_stream_intel_);
 }
 
-envoy_error Client::DirectStreamCallbacks::streamError() {
+void Client::DirectStreamCallbacks::latchError() {
+  if (error_.has_value()) {
+  }
   envoy_error error{};
+  error_ = error;
+
   OptRef<RequestDecoder> request_decoder = direct_stream_.requestDecoder();
   if (!request_decoder) {
-    return error;
+    error_->message = envoy_nodata;
+    return;
   }
   const auto& info = request_decoder->streamInfo();
 
   if (info.responseCode().has_value()) {
-    error.error_code = Bridge::Utility::errorCodeFromLocalStatus(
+    error_->error_code = Bridge::Utility::errorCodeFromLocalStatus(
         static_cast<Http::Code>(info.responseCode().value()));
   } else if (StreamInfo::isStreamIdleTimeout(info)) {
-    error.error_code = ENVOY_REQUEST_TIMEOUT;
+    error_->error_code = ENVOY_REQUEST_TIMEOUT;
   } else {
-    error.error_code = ENVOY_STREAM_RESET;
+    error_->error_code = ENVOY_STREAM_RESET;
   }
 
   if (info.responseCodeDetails().has_value()) {
-    error.message = Data::Utility::copyToBridgeData(info.responseCodeDetails().value());
+    error_->message = Data::Utility::copyToBridgeData(info.responseCodeDetails().value());
   } else {
-    error.message = envoy_nodata;
+    error_->message = envoy_nodata;
   }
 
-  error.attempt_count = info.attemptCount().value_or(0);
-  return error;
+  error_->attempt_count = info.attemptCount().value_or(0);
 }
 
 Client::DirectStream::DirectStream(envoy_stream_t stream_handle, Client& http_client)
     : stream_handle_(stream_handle), parent_(http_client) {}
 
 Client::DirectStream::~DirectStream() { ENVOY_LOG(debug, "[S{}] destroy stream", stream_handle_); }
 
+CodecEventCallbacks*
+Client::DirectStream::registerCodecEventCallbacks(CodecEventCallbacks* codec_callbacks) {
+  // registerCodecEventCallbacks is called with nullptr when the underlying
+  // Envoy stream is going away. Make sure Envoy Mobile sees the stream as
+  // closed as well.
+  if (codec_callbacks == nullptr &&
```

Comment:
As discussed with Ryan this triggered in some java test of "create a zillion connections and close half" which triggered the "this looks sketch like frameshift"
filed https://github.com/envoyproxy/envoy/issues/31966 for this but there's other cases where Envoy does close the connection and we need this error handling so I'll 
1) add C++ tests which explicitly turn up and test frameshift to make sure we cover this
2) turn off frameshift by default for E-M